Encoder/DefaultEncoder.php

```diff
@@ -0,0 +1,163 @@
+     * @var string
+     */
+    protected $encryptionEngine;
+
+    /**
+     * @var string
+     */
+    protected $encryptionAlgorithm;
+
+    /**
+     * Constructor.
+     *
+     * @param KeyLoaderInterface $keyLoader
+     * @param string             $encryptionEngine
+     * @param string             $encryptionAlgorithm
+     *
+     * @throws InvalidArgumentException If the given encryptionEngine is not supported.
+     * @throws InvalidArgumentException If the given algorithm is not supported.
+     */
+    public function __construct(KeyLoaderInterface $keyLoader, $encryptionEngine, $encryptionAlgorithm)
+    {
+        if (!in_array($encryptionEngine, ['OpenSSL', 'SecLib'])) {
+            throw new InvalidArgumentException(
+                sprintf('The encryption engine "%s" is not supported', $encryptionEngine)
+            );
+        }
+
+        if (!$this->isAlgorithmSupportedForEngine($encryptionEngine, $encryptionAlgorithm)) {
+            throw new InvalidArgumentException(
+                sprintf('The algorithm "%s" is not supported for %s', $encryptionAlgorithm, $encryptionEngine)
+            );
+        }
+
+        $this->keyLoader           = $keyLoader;
+        $this->encryptionEngine    = $encryptionEngine;
+        $this->encryptionAlgorithm = $encryptionAlgorithm;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \RuntimeException If the JWS cannot be signed
+     */
+    public function encode(array $data)
+    {
+        $jws = new JWS(['alg' => $this->encryptionAlgorithm], $this->encryptionEngine);
+        $key = $this->keyLoader->loadKey('private');
+
+        $jws->setPayload($data);
+
+        try {
+            $jws->sign($key, $this->keyLoader->getPassphrase());
+        } catch (InvalidArgumentException $e) {
+            throw new \RuntimeException(
+                sprintf('An error occurred while signing the JWS (%s). Please check your configuration.', $e->getMessage())
+            );
+        }
+
+        return $jws->getTokenString();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws InvalidArgumentException If the JWS cannot be loaded
+     * @throws \RuntimeException        If the JWS cannot be verified
+     */
+    public function decode($token)
+    {
+        try {
+            $jws = JWS::load($token, false, null, $this->encryptionEngine);
+        } catch (InvalidArgumentException $e) {
+            return false;
+        }
+
+        $key     = $this->keyLoader->loadKey('public');
+        $payload = $jws->getPayload();
+
+        try {
+            if (!$jws->verify($key, $this->encryptionAlgorithm) || $this->isJWSExpired($payload)) {
+                return false;
+            }
+        } catch (InvalidArgumentException $e) {
+            throw new \RuntimeException(
+                sprintf('An error occurred while verifying the JWS (%s). Please check your configuration.', $e->getMessage())
+            );
+        }
+
+        return $payload;
+    }
```

Comment:
There's a slight problem with this method - it exists in the current implementation too. There are 3 big reasons why this method might fail:

a) the token is invalid - it literally can't be decoded
b) the token is expired
c) the token's signature failed

But, there's no way to know which occurred - all return false. What do you think about throwing an exception instead of returning false? We could have 3 exceptions, each which extend the same base exception class (something like `TokenDecodeFailureException`. This is similar to how Guzzle behaves: it throws an exception when a request fails (and has different exception classes for the different failure types)